### PR TITLE
Remove Helper Method Frames (HMF) from Reflection

### PR DESCRIPTION
**Pull request by AaronRobinsonMSFT**
Convert `Signature.GetSignature()` to QCall.
Rename `Signature.GetSignature()` to `Signature.Init()`.
Remove HMF from `Signature.GetTypeParameterOffset()`.
Remove HMF from `Signature.GetCallingConventionFromFunctionPointerAtOffset()`.
Convert `Signature.GetCustomModifiersAtOffset()` to QCall.
Convert `RuntimeTypeHandle.GetDeclaringMethodForGenericParameter()` to QCall.
Rename `RuntimeTypeHandle.GetDeclaringMethod()` to `RuntimeTypeHandle.GetDeclaringMethodForGenericParameter()`.
Convert `RuntimeTypeHandle.CanCastTo()` to slow/fast paths using FCall/QCall.
Remove HMF from `RuntimeMethodHandle::GetMethodDef()`.
Convert `RuntimeMethodHandle.GetStubIfNeeded()` to fast/slow paths using FCall/QCall.
Convert `RuntimeTypeHandle.SatisfiesConstraints()` to QCall.
Convert `RuntimeMethodHandle.GetMethodBody()` to QCall.

### CODE REVIEW

File: src/coreclr/vm/runtimehandles.h [hunk 1]
```diff
@@ -59,7 +59,7 @@ class RuntimeExceptionHandlingClause : Object
     INT32 _filterOffset;
 };
 
-class RuntimeMethodBody : Object
+class RuntimeMethodBody : public Object
```
Comment (jkotas) [inline src/coreclr/vm/runtimehandles.h:62]: The other managed types defined in this file can use the same treatment.

**Comment by am11** [comment]
S.D.Process test failure is unrelated https://github.com/dotnet/runtime/issues/110643 (intermittently failing on other PRs as well).

File: src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs [hunk 2]
```diff
@@ -1895,68 +1958,94 @@ internal static int GetMDStreamVersion(RuntimeModule module)
...
+
+        [MemberNotNull(nameof(_declaringType))]
+        [MemberNotNull(nameof(_returnTypeORfieldType))]
+        private void Init(
+            void* pCorSig, int cCorSig,
+            RuntimeFieldHandleInternal fieldHandle,
+            RuntimeMethodHandleInternal methodHandle)
+        {
+            Signature _this = this;
+            IntPtr typeHandle = _declaringType?.GetUnderlyingNativeHandle() ?? IntPtr.Zero;
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs:1996]: Can the `_declaringType` initialization be made more sensible by always assigning it on the managed side before Init is called?

**Comment by jkotas** [comment]
> Copilot reviewed 5 out of 19 changed files in this pull request and generated 1 comment.

I wanted to give it a try to see what it is going to say on this PR. No actionable feedback.

File: src/coreclr/System.Private.CoreLib/src/ILLink/ILLink.Descriptors.Shared.xml [hunk 3]
```diff
@@ -9,6 +9,9 @@
     <!-- Accessed via native code. -->
     <type fullname="System.Runtime.InteropServices.CustomMarshalers.*" />
 
+    <!-- Share definition across managed and native code. -->
```
[the remark this answers is not in the document]
Comment (AaronRobinsonMSFT) [reply src/coreclr/System.Private.CoreLib/src/ILLink/ILLink.Descriptors.Shared.xml:12]: > I do not understand this comment. It should not be necessary to root the type just to share the layout with the native VM code.

The issue is the trimmer sees two fields that are unset and unused and therefore there are removed. They are set in unmanaged code, but the type needs to have the same layout in both. The layout doesn't need to be defined in `corelib.h` because it is directly defined in unmanaged, so `corelib.h` has no notion of `Signature`.

> Should this use corelib.h instead?

Perhaps, but it is going to introduce yet another description of the type's fields - managed, unmanaged and `corelib.h`. The definitions in `corelib.h` will never be used. And of course if someone does remove them then things break. I prefer to make the retention in the C# or XML side because that is clear from a trimmer case. The `corelib.h` file is simply a crutch.

[the remark this answers is not in the document]
Comment (jkotas) [reply src/coreclr/System.Private.CoreLib/src/ILLink/ILLink.Descriptors.Shared.xml:12]: This is covered in https://github.com/dotnet/runtime/blob/main/docs/design/coreclr/botr/corelib.md#types-with-a-managedunmanaged-duality. `DEFINE_CLASS_U` and `DEFINE_FIELD_U` macros in `corelib.h` are meant to be used for situations where we have managed and unmanaged layouts that are expected to match. Checked builds will assert when the type size or the field offsets do not match between the two layouts.

The [CreateRuntimeRootDescriptorFile task](https://github.com/dotnet/runtime/blob/3aa1ec5bb1f50f0a1bed9cfcac8734f742bcf24b/src/tools/illink/src/ILLink.Tasks/CreateRuntimeRootDescriptorFile.cs#L223-L231) understands `DEFINE_CLASS_U` and preserves all fields for types with managed/unmanaged duality.

File: src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs [hunk 4]
```diff
@@ -660,7 +660,24 @@ internal static unsafe MdUtf8String GetUtf8Name(RuntimeType type)
         }
 
         [MethodImpl(MethodImplOptions.InternalCall)]
-        internal static extern bool CanCastTo(RuntimeType type, RuntimeType target);
+        private static extern CastResult CanCastToInternal(RuntimeType type, RuntimeType target);
```
[the remark this answers is not in the document]
Comment (AaronRobinsonMSFT) [reply src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs:663]: Done.

File: src/coreclr/vm/runtimehandles.cpp [hunk 5]
```diff
@@ -1471,242 +1452,194 @@ FCIMPL3(INT32, SignatureNative::GetParameterOffsetInternal, PCCOR_SIGNATURE sig,
...
-    if (!pMethodUNSAFE)
-        FCThrowRes(kArgumentNullException, W("Arg_InvalidHandle"));
-
-    MethodDesc* pMethod = pMethodUNSAFE->GetMethod();
+FCIMPL1(INT32, RuntimeMethodHandle::GetMethodDef, MethodDesc* pMethod)
+{
+    FCALL_CONTRACT;
+    _ASSERTE(pMethod != NULL);
 
     if (pMethod->HasMethodInstantiation())
```
[the remark this answers is not in the document]
Comment (AaronRobinsonMSFT) [reply src/coreclr/vm/runtimehandles.cpp:1641]: Agree.

File: src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs (hunk 2 above)
Comment (AaronRobinsonMSFT) [reply src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs:1996]: Oh, right. We have `RuntimeMethodHandle.GetDeclaringType()`. Thanks.

File: src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (jkotas) [reply src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs:663]: Also, the FCall should be deleted and `CastCache.TryGet` called instead.

File: src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs [hunk 6]
```diff
@@ -819,23 +858,17 @@ internal bool ContainsGenericVariables()
             return ContainsGenericVariables(GetRuntimeTypeChecked());
         }
 
-        [MethodImpl(MethodImplOptions.InternalCall)]
-        private static extern bool SatisfiesConstraints(RuntimeType paramType, IntPtr* pTypeContext, int typeContextLength, IntPtr* pMethodContext, int methodContextLength, RuntimeType toType);
+        [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "RuntimeTypeHandle_SatisfiesConstraints")]
+        private static partial Interop.BOOL SatisfiesConstraints(QCallTypeHandle paramType, nint pTypeContext, RuntimeMethodHandleInternal pMethodContext, QCallTypeHandle toType);
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs:862]: Nit: `pTypeContext` can be `QCallTypeHandle`. Using two different styles to pass a type in a signature does not look right.

File: src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs [hunk 7]
```diff
@@ -1267,8 +1328,17 @@ internal static IRuntimeMethodInfo StripMethodInstantiation(IRuntimeMethodInfo m
         [MethodImpl(MethodImplOptions.InternalCall)]
         internal static extern Resolver GetResolver(RuntimeMethodHandleInternal method);
 
-        [MethodImpl(MethodImplOptions.InternalCall)]
-        internal static extern RuntimeMethodBody? GetMethodBody(IRuntimeMethodInfo method, RuntimeType declaringType);
+        [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "RuntimeMethodHandle_GetMethodBody")]
+        private static partial void GetMethodBody(RuntimeMethodHandleInternal method, nint declaringType, ObjectHandleOnStack result);
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs:1332]: `declaringType` can be `QCallTypeHandle`

File: src/coreclr/vm/runtimehandles.cpp [hunk 8]
```diff
@@ -1471,362 +1453,290 @@ FCIMPL3(INT32, SignatureNative::GetParameterOffsetInternal, PCCOR_SIGNATURE sig,
...
+    if (callConv == IMAGE_CEE_CS_CALLCONV_FIELD)
+    {
+        msig.NextArgNormalized();
 
-            OBJECTREF refRetType = msig.GetRetTypeHandleThrowing().GetManagedClassObject();
-            gc.pSig->SetReturnType(refRetType);
+        OBJECTREF refRetType = msig.GetLastTypeHandleThrowing().GetManagedClassObject();
+        gc.pSig->SetReturnType(refRetType);
+    }
+    else if (pMethodDesc == NULL || !pMethodDesc->IsDynamicMethod())
```
Comment (jkotas) [inline src/coreclr/vm/runtimehandles.cpp:1714]:
```suggestion
    else
```
I think `IsDynamicMethod` condition is hard to follow. We can skip more code with easier to understand condition - see my suggestion above.

File: src/coreclr/vm/runtimehandles.cpp [hunk 9]
```diff
@@ -1471,362 +1453,290 @@ FCIMPL3(INT32, SignatureNative::GetParameterOffsetInternal, PCCOR_SIGNATURE sig,
...
-
-        PREFIX_ASSUME(pCorSig!= NULL);
-        BYTE callConv = *(BYTE*)pCorSig;
-        SigTypeContext typeContext;
-        if (pMethod)
-            SigTypeContext::InitTypeContext(
-                pMethod, declType.GetClassOrArrayInstantiation(), pMethod->LoadMethodInstantiation(), &typeContext);
-        else
-            SigTypeContext::InitTypeContext(declType, &typeContext);
+    uint32_t callConv;
```
Comment (jkotas) [inline src/coreclr/vm/runtimehandles.cpp:1688]:
```suggestion
    // Initialize _returnTypeORfieldType and _arguments if it was not initialized yet
    if (gc.pSig->m_returnType != NULL)
    {
        uint32_t callConv;
```

File: src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs [hunk 10]
```diff
@@ -1895,68 +1965,96 @@ internal static int GetMDStreamVersion(RuntimeModule module)
...
+        private RuntimeMethodHandleInternal _pMethod;
         #endregion
 
+        [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "Signature_Init")]
+        private static partial void Init(
+            ObjectHandleOnStack _this,
+            void* pCorSig, int cCorSig,
+            RuntimeFieldHandleInternal fieldHandle,
+            RuntimeMethodHandleInternal methodHandle,
+            IntPtr typeHandle);
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs:1993]: `IntPtr typeHandle` argument is unnecessary (the type can be read from `_this` on the unmanaged side).

File: src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs [hunk 11]
```diff
@@ -1895,68 +1965,96 @@ internal static int GetMDStreamVersion(RuntimeModule module)
...
         }
         #endregion
 
         #region Internal Members
-        internal CallingConventions CallingConvention => (CallingConventions)(byte)m_managedCallingConventionAndArgIteratorFlags;
-        internal RuntimeType[] Arguments => m_arguments;
-        internal RuntimeType ReturnType => m_returnTypeORfieldType;
-        internal RuntimeType FieldType => m_returnTypeORfieldType;
+        internal CallingConventions CallingConvention => (CallingConventions)(_managedCallingConventionAndArgIteratorFlags & 0xff);
+        internal RuntimeType[] Arguments => _arguments ?? [];
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs:2055]: Should we rather assert that  `arguments` is non-null here?

**Review by jkotas — APPROVED** [review]
LGTM otherwise. Thanks!

File: src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs [hunk 12]
```diff
@@ -1895,68 +1965,96 @@ internal static int GetMDStreamVersion(RuntimeModule module)
...
-        internal RuntimeType[] m_arguments;
-        internal RuntimeType m_declaringType;
-        internal RuntimeType m_returnTypeORfieldType;
-        internal object? m_keepalive;
-        internal void* m_sig;
-        internal int m_managedCallingConventionAndArgIteratorFlags; // lowest byte is CallingConvention, upper 3 bytes are ArgIterator flags
-        internal int m_nSizeOfArgStack;
-        internal int m_csig;
-        internal RuntimeMethodHandleInternal m_pMethod;
+        private RuntimeType[]? _arguments;
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs:1972]: It may be nice to rename the fields in the unmanaged view to match.

File: src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs (hunk 11 above)
Comment (AaronRobinsonMSFT) [reply src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs:2055]: > It should be a bug if Arguments property is called on a field signature.

Right. Should I throw an `InvalidOperationException` instead? I'm also fine just asserting, but I'm not sure the trade off between assert vs exception when we really don't expect the case.

Comment (jkotas) [reply src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs:2055]: This is internal property. Assert should be fine. It is good enough to assert to catch bugs in our code.

File: src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs (hunk 6 above)
Comment (AaronRobinsonMSFT) [reply src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs:862]: In this case it can't without changing `QCallTypeHandle` to accept `RuntimeType?`, that is why it is `IntPtr`. I'm fine changing it to that if that is the preference.

**Comment by AaronRobinsonMSFT** [comment]
@jkotas I did some consolidation around `TypeHandle.CanCastTo`, please take another look.

File: src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs (hunk 6 above)
Comment (AaronRobinsonMSFT) [reply src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs:848]: Oh, I see. The constructor is overly restrictive for no reason. The impl already handles the case. Will fix.

**Comment by jkotas** [comment]
> I did some consolidation around TypeHandle.CanCastTo, please take another look.

I think we can go a bit further and merge the QCALL for the regular no-cache version and reflection no-cache version by moving the Nullable<T> special case to the managed side. It should make the casting impl details more centralized. I left a suggestion for what it may look like.

**Comment by AaronRobinsonMSFT** [comment]
> > I did some consolidation around TypeHandle.CanCastTo, please take another look.
> 
> I think we can go a bit further and merge the QCALL for the regular no-cache version and reflection no-cache version by moving the Nullable special case to the managed side. It should make the casting impl details more centralized. I left a suggestion for what it may look like.

Good call, simplified a lot.

File: src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs [hunk 13]
```diff
@@ -1098,28 +1098,60 @@ public static TypeHandle TypeHandleOf<T>()
...
         public bool CanCastTo(TypeHandle destTH)
         {
-            if (m_asTAddr == destTH.m_asTAddr)
-                return true;
+            return TryCanCastTo(this, destTH) switch
+            {
+                CastResult.CanCast => true,
+                CastResult.CannotCast => false,
+                _ => CanCastTo_NoCacheLookup(m_asTAddr, destTH.m_asTAddr) != Interop.BOOL.FALSE
+            };
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs:1111]:
```suggestion
                _ => CanCastToWorker(m_asTAddr, destTH.m_asTAddr);
            };
            
            [MethodImpl(MethodImplOptions.NoInlining)]
            static bool CanCastToWorker(TypeHandle srcTH, TypeHandle destTH)
            {
                 // Regular casting does not allow T to be cast to Nullable<T>
                 // see TypeHandle::CanCastTo and ObjIsInstanceOfCore
                 if (!srcTH.IsTypeDesc
                    && !destTH.IsTypeDesc
                    && CastHelpers.IsNullableForType(destTH.AsMethodTable(), srcTH.AsMethodTable()))
                 {
                     return false;
                 }

                 return CanCastTo_NoCacheLookup(srcTH.m_asTAddr, destTH.m_asTAddr) != Interop.BOOL.FALSE;
            }
```
I think this needs to have the handling of the nullable special case inlined as well.

**Review by jkotas — APPROVED** [review]
LGTM!

**Comment by jkotas** [comment]
@EgorBot -windows_intel
```csharp
using System;
using BenchmarkDotNet.Attributes;

public class Bench
{
    Type fromType = typeof(object);
    Type toType = typeof(string);

    [Benchmark]
    public bool IsAssignableFrom() => toType.IsAssignableFrom(fromType);
}
```

**Comment by AaronRobinsonMSFT** [comment]
Seems the change is net zero for `IsAssignableFrom` - https://github.com/EgorBot/runtime-utils/issues/205#issuecomment-2547397344.

**Comment by jkotas** [comment]
@EgorBot -windows_intel
```csharp
using System;
using BenchmarkDotNet.Attributes;

public class Bench
{
    Type fromType = typeof(string);
    Type toType = typeof(object);

    [Benchmark]
    public bool IsAssignableFrom() => toType.IsAssignableFrom(fromType);
}
```
